exercises/wordy/src/example.c

```diff
@@ -0,0 +1,174 @@
+#include "wordy.h"
+
+#define ADDITION 1
+#define SUBRACTION 2
+#define DIVISION 3
+#define MULTIPLICATION 4
+#define EXPO 5
+
+static char *strdup(const char *str)
+{
+   char *copy;
+   if ((copy = calloc(strlen(str) + 1, sizeof(char *))) != NULL) {
+      strcpy(copy, str);
+   }
+   return copy;
+}
+
+static char *strsep(char **stringp, const char *delim)
+{
+   char *s;
+   const char *spanp;
+   int c, sc;
+   char *tok;
+   if ((s = *stringp) == NULL)
+      return (NULL);
+   for (tok = s;;) {
+      c = *s++;
+      spanp = delim;
+      do {
+         if ((sc = *spanp++) == c) {
+            if (c == 0)
+               s = NULL;
+            else
+               s[-1] = 0;
+            *stringp = s;
+            return (tok);
+         }
+      } while (sc != 0);
+   }
+}
+
+static int raise_to(int base, int exp)
+{
+   int ret = 1;
+   for (int i = 0; i < exp; i++) {
+      ret = ret * base;
+   }
+   return ret;
+}
+
+static void free_list(char **list, size_t n)
+{
+   for (size_t i = 0; i < n; ++i) {
+      free(list[i]);
+   }
+   free(list);
+}
+
+static char *is_integer(char *s)
+{
+   size_t s_len = strlen(s);
+   size_t i = 0;
+   if (strstr(s, "-")) {
+      i = 1;
+   }
+   for (; i < s_len; i++) {
+      if (!isdigit(s[i])) {
+         return NULL;
+      }
+   }
+   return s;
+}
+
+static char **create_list(const char *input, size_t * n)
+{
+   char *s = strdup(input);
+   size_t list_alloc = 1;
+   size_t list_used = 0;
+   char **lists = calloc(list_alloc, sizeof(char *));
+   char *list, *rest = s;
+   while ((list = strsep(&rest, " ")) != NULL) {
+      if (list_used == list_alloc) {
+         list_alloc *= 2;
+         lists = realloc(lists, list_alloc * sizeof(char *));
+      }
+      lists[list_used++] = strdup(list);
+   }
+   if (list_used == 0) {
+      free(lists);
+      lists = NULL;
+   } else {
+      lists = realloc(lists, list_used * sizeof(char *));
+   }
+   *n = list_used;
+   free(s);
+   return lists;
+}
+
+int wordy(const char *instructions)
+{
+   int result = 0;
+   int operator = - 1;
+   int no_op = 1;
+   char *buf1 = NULL;
+   char *buf2 = NULL;
+   size_t n = 0;
+   size_t instructions_len = strlen(instructions);
+   if (instructions[instructions_len - 1] != '?') {
+      return result;
+   }
+   char *input = calloc(instructions_len, sizeof(char *));
+   strncpy(input, instructions, instructions_len - 1);
+   char **l = create_list(input, &n);
+   free(input);
+   if (strcmp(l[0], "What") || strcmp(l[1], "is") || n == 0) {
+      free_list(l, n);
+      return result;
+   }
+   for (size_t i = 2; i < n; ++i) {
+      if ((buf1 = is_integer(l[i])) != NULL && i == 2) {
+         result = atoi(buf1);
+      }
+      if (((buf1 = is_integer(l[i])) != NULL && i != 2)
+          || (strstr(l[i], "th") && i != 2 && isdigit(l[i][0]))) {
+         if (operator == ADDITION) {
```

Comment:
yes , the parent loop never enters the code block until first digit is detected in the first item in the list because of the syntax , Always "What is x plus y?" or "What is x plus y multiplied by z divided by 1?" , so when the string is divided into list , "What" , "is" , x , "plus" , y , "......" . by this x must be the second item in the list , if not it will be a syntax error. and the third item must be a operator so the operator will be set right after x is detected if not it results in syntax error , all the test ran successfully